source/common/http/filter_manager.cc

```diff
@@ -1264,6 +1266,12 @@ bool ActiveStreamDecoderFilter::recreateStream() {
   parent_.stream_info_.setResponseCodeDetails(
       StreamInfo::ResponseCodeDetails::get().InternalRedirect);
 
+  if (headers != nullptr) {
+    ResponseHeaderMapPtr headers_copy = createHeaderMap<ResponseHeaderMapImpl>(*headers);
+    parent_.filter_manager_callbacks_.setResponseHeaders(std::move(headers_copy));
```

Comment:
The headers are not used directly here; I make a copy of the header map (createHeaderMap copies the object itself) and use the reference to that instead, so the original pointer will not be invalidated. 

setResponseHeaders is needed for the current headers to be logged in the access logs. Without this, the headers are overwritten when the stream is destroyed. The "test-header" logic in the redirect_integration_test tests this behavior; without this call, that header is not available to the logs.

Reply:
Looking more closely at Filter::onUpstreamHeaders, it seems that the headers passed in to setupRedirect are moved into encodeHeaders when setupRedirect does not happen or fails.  I think that what you're doing is mostly fine, but wonder if it would be possible for the call to setResponseHeaders to happen in Filter::onUpstreamHeaders before calling into either encodeHeaders or setupRedirect.  I don't think I have enough understanding of the surrounding code to really advice, so I guess your version is fine.  See if senior maintainers have some further advice for you.

Reply:
Since the call to setResponseHeaders takes ownership of the headers, it can't happen earlier in Filter::onUpstreamHeaders. We don't know until setupRedirect returns whether that operation succeeded; if it failed, then headers will continue to be modified and then std::move'd to encodeHeaders, which would have undefined behavior if they were already std::move'd. Calling it after recreateStream would set them too late, and they won't be properly logged. It's not really feasible to move the call without either preemptively making a copy of the headers, or greatly restructuring the code (maybe).